hazelcast/src/main/java/com/hazelcast/sql/impl/SqlInternalService.java

```diff
@@ -0,0 +1,194 @@
+        stateRegistry = new QueryStateRegistry(nodeServiceProvider);
+
+        // Operation handler depends on state registry.
+        operationHandler = new QueryOperationHandlerImpl(
+            instanceName,
+            nodeServiceProvider,
+            serializationService,
+            stateRegistry,
+            outboxBatchSize,
+            FLOW_CONTROL_FACTORY,
+            fragmentThreadCount,
+            operationThreadCount
+        );
+
+        // State checker depends on state registries and operation handler.
+        stateRegistryUpdater = new QueryStateRegistryUpdater(
+            instanceName,
+            nodeServiceProvider,
+            stateRegistry,
+            operationHandler,
+            stateCheckFrequency
+        );
+    }
+
+    public void start() {
+        stateRegistryUpdater.start();
+    }
+
+    public void reset() {
+        stateRegistry.reset();
+    }
+
+    public void shutdown() {
+        stateRegistryUpdater.stop();
+        operationHandler.stop();
+
+        reset();
+    }
+
+    /**
+     * Internal query execution routine.
+     *
+     * @return Query state.
+     */
+    public QueryState execute(Plan plan, List<Object> params, long timeout, int pageSize) {
+        // Get local member ID and check if it is still part of the plan.
+        UUID localMemberId = nodeServiceProvider.getLocalMemberId();
+
+        if (!plan.getPartitionMap().containsKey(localMemberId)) {
+            throw QueryException.memberLeave(localMemberId);
+        }
+
+        // Prepare mappings.
+        QueryExecuteOperationFactory operationFactory = new QueryExecuteOperationFactory(
+            plan,
+            params,
+            createEdgeInitialMemoryMapForPlan(plan)
+        );
+
+        // Register the state.
+        BlockingRootResultConsumer consumer = new BlockingRootResultConsumer();
+
+        QueryState state = stateRegistry.onInitiatorQueryStarted(
+            localMemberId,
+            timeout,
+            plan,
+            consumer,
+            operationHandler,
+            true
+        );
+
+        try {
+            // Start execution on local member.
+            QueryExecuteOperation localOp = operationFactory.create(state.getQueryId(), localMemberId);
+
+            localOp.setRootConsumer(consumer, pageSize);
+
+            operationHandler.submitLocal(localMemberId, localOp);
+
+            // Start execution on remote members.
+            for (UUID memberId : plan.getMemberIds()) {
+                if (memberId.equals(localMemberId)) {
+                    continue;
+                }
+
+                QueryExecuteOperation remoteOp = operationFactory.create(state.getQueryId(), memberId);
+
+                if (!operationHandler.submit(localMemberId, memberId, remoteOp)) {
+                    throw QueryException.memberConnection(memberId);
+                }
+            }
+
+            return state;
+        } catch (Exception e) {
+            state.cancel(e);
```

Comment:
Shall we log the exception here?

Reply:
The exception will be propagated to the user through API, so there is no serious need to log it.